File: game/ai/policy_guard.cpp
#include "game/ai/policy_guard.h"

#include <algorithm>
#include <chrono>
#include <ctime>
#include <fstream>
#include <iomanip>
#include <sstream>

namespace game::ai {
namespace {

std::string Trim(const std::string& s) {
  const auto start = s.find_first_not_of(" \t\r\n");
  if (start == std::string::npos) {
    return "";
  }
  const auto end = s.find_last_not_of(" \t\r\n");
  return s.substr(start, end - start + 1);
}

bool StartsWith(const std::string& value, const std::string& prefix) {
  return value.size() >= prefix.size() &&
         value.compare(0, prefix.size(), prefix) == 0;
}

std::string ToIso8601Now() {
  const auto now = std::chrono::system_clock::now();
  const std::time_t now_c = std::chrono::system_clock::to_time_t(now);
  std::tm tm{};
#if defined(_WIN32)
  gmtime_s(&tm, &now_c);
#else
  gmtime_r(&now_c, &tm);
#endif
  std::ostringstream out;
  out << std::put_time(&tm, "%Y-%m-%dT%H:%M:%SZ");
  return out.str();
}

}  // namespace

PolicyGuard::PolicyGuard(std::filesystem::path policy_path,
                         std::filesystem::path audit_log_path)
    : policy_path_(std::move(policy_path)),
      audit_log_path_(std::move(audit_log_path)) {
  ReloadPolicy();
}

bool PolicyGuard::ReloadPolicy() {
  std::ifstream in(policy_path_);
  if (!in.is_open()) {
    return false;
  }

  policy_ = ExecutionPolicy{};

  enum class Context { None, AllowWritable, DenyWritable, AllowDomain, AllowPorts };

  Context context = Context::None;
  bool in_ai_writable = false;
  bool in_network_allow = false;

  std::string line;
  while (std::getline(in, line)) {
    const auto trimmed = Trim(line);
    if (trimmed.empty() || StartsWith(trimmed, "#")) {
      continue;
    }

    if (trimmed == "autonomy:" || trimmed == "network:" ||
        trimmed == "resource_limits:" || trimmed == "cpu:" ||
        trimmed == "memory:" || trimmed == "execution:" ||
        trimmed == "circuit_breakers:") {
      context = Context::None;
      in_ai_writable = false;
      if (trimmed != "network:") {
        in_network_allow = false;
      }
      continue;
    }

    if (trimmed == "ai_writable:") {
      in_ai_writable = true;
      context = Context::None;
      continue;
    }

    if (trimmed == "allow:") {
      if (in_ai_writable) {
        context = Context::AllowWritable;
      }
      if (!in_ai_writable) {
        in_network_allow = true;
      }
      continue;
    }

    if (trimmed == "deny:") {
      if (in_ai_writable) {
        context = Context::DenyWritable;
      }
      if (!in_ai_writable) {
        in_network_allow = false;
      }
      continue;
    }

    if (trimmed == "domains:") {
      context = in_network_allow ? Context::AllowDomain : Context::None;
      continue;
    }

    if (trimmed == "ports:") {
      context = in_network_allow ? Context::AllowPorts : Context::None;
      continue;
    }

    if (StartsWith(trimmed, "- ")) {
      const auto item = Trim(trimmed.substr(2));
      switch (context) {
        case Context::AllowWritable:
          policy_.writable_allow_prefixes.push_back(item);
          break;
        case Context::DenyWritable:
          policy_.writable_deny_prefixes.push_back(item);
          break;
        case Context::AllowDomain:
          if (!item.empty() && item != "\"*\"") {
            policy_.allowed_domains.push_back(item);
          }
          break;
        case Context::AllowPorts:
          policy_.allowed_ports.push_back(std::stoi(item));
          break;
        default:
          break;
      }
      continue;
    }

    if (StartsWith(trimmed, "max_percent:")) {
      policy_.limits.max_cpu_percent =
          std::stoi(Trim(trimmed.substr(std::string("max_percent:").size())));
    } else if (StartsWith(trimmed, "max_ram_mb:")) {
      policy_.limits.max_ram_mb =
          std::stoi(Trim(trimmed.substr(std::string("max_ram_mb:").size())));
    } else if (StartsWith(trimmed, "max_time_seconds:")) {
      policy_.limits.max_time_seconds =
          std::stoi(Trim(trimmed.substr(std::string("max_time_seconds:").size())));
    } else if (StartsWith(trimmed, "max_failed_deployments:")) {
      policy_.circuit_breakers.max_failed_deployments = std::stoi(
          Trim(trimmed.substr(std::string("max_failed_deployments:").size())));
    } else if (StartsWith(trimmed, "max_regression_threshold:")) {
      policy_.circuit_breakers.max_regression_threshold = std::stod(
          Trim(trimmed.substr(std::string("max_regression_threshold:").size())));
    } else if (StartsWith(trimmed, "emergency_disable_file:")) {
      policy_.circuit_breakers.emergency_disable_file =
          Trim(trimmed.substr(std::string("emergency_disable_file:").size()));
    }
  }

  if (policy_.writable_allow_prefixes.empty()) {
    policy_.writable_allow_prefixes = {"game/ai/", "policies/", "logs/", "tools/"};
  }

  return true;
}

DeploymentDecision PolicyGuard::EnforceBeforePatchDeployment(
    const std::vector<std::filesystem::path>& touched_files,
    const std::optional<std::string>& outbound_domain,
    int outbound_port,
    int requested_cpu_percent,
    int requested_ram_mb,
    int requested_runtime_seconds,
    double regression_score) const {
  std::string reason;
  for (const auto& path : touched_files) {
    if (!IsWritablePathAllowed(path, &reason)) {
      return {false, reason};
    }
  }

  if (!IsNetworkAllowed(outbound_domain, outbound_port, &reason)) {
    return {false, reason};
  }

  if (!AreResourcesAllowed(requested_cpu_percent, requested_ram_mb,
                           requested_runtime_seconds, &reason)) {
    return {false, reason};
  }

  if (IsCircuitBreakerOpen(regression_score, &reason)) {
    return {false, reason};
  }

  return {true, "allowed"};
}

void PolicyGuard::RecordProposed(const std::string& change_id,
                                 const std::string& summary) const {
  Audit(ChangeAction::Proposed, change_id, {{"summary", summary}});
}

void PolicyGuard::RecordApplied(const std::string& change_id,
                                const std::string& summary,
                                bool success) const {
  Audit(ChangeAction::Applied, change_id,
        {{"summary", summary}, {"success", success ? "true" : "false"}});
}

void PolicyGuard::RecordReverted(const std::string& change_id,
                                 const std::string& summary,
                                 const std::string& reason) const {
  Audit(ChangeAction::Reverted, change_id,
        {{"summary", summary}, {"reason", reason}});
}

bool PolicyGuard::IsWritablePathAllowed(const std::filesystem::path& path,
                                        std::string* reason) const {
  const auto path_str = path.generic_string();

  for (const auto& deny : policy_.writable_deny_prefixes) {
    if (StartsWith(path_str, deny)) {
      if (reason) {
        *reason = "write denied for path: " + path_str;
      }
      return false;
    }
  }

  for (const auto& allow : policy_.writable_allow_prefixes) {
    if (StartsWith(path_str, allow)) {
      return true;
    }
  }

  if (reason) {
    *reason = "write outside allowed scope: " + path_str;
  }
  return false;
}

bool PolicyGuard::IsNetworkAllowed(const std::optional<std::string>& domain,
                                   int port,
                                   std::string* reason) const {
  if (!domain.has_value()) {
    return true;
  }

  if (std::find(policy_.allowed_domains.begin(), policy_.allowed_domains.end(),
                domain.value()) == policy_.allowed_domains.end()) {
    if (reason) {
      *reason = "domain not allowed: " + domain.value();
    }
    return false;
  }

  if (std::find(policy_.allowed_ports.begin(), policy_.allowed_ports.end(), port) ==
      policy_.allowed_ports.end()) {
    if (reason) {
      *reason = "port not allowed: " + std::to_string(port);
    }
    return false;
  }

  return true;
}

bool PolicyGuard::AreResourcesAllowed(int cpu_percent,
                                      int ram_mb,
                                      int runtime_seconds,
                                      std::string* reason) const {
  if (cpu_percent > policy_.limits.max_cpu_percent) {
    if (reason) {
      *reason = "cpu request exceeds max policy";
    }
    return false;
  }

  if (ram_mb > policy_.limits.max_ram_mb) {
    if (reason) {
      *reason = "ram request exceeds max policy";
    }
    return false;
  }

  if (runtime_seconds > policy_.limits.max_time_seconds) {
    if (reason) {
      *reason = "runtime request exceeds max policy";
    }
    return false;
  }

  return true;
}

bool PolicyGuard::IsCircuitBreakerOpen(double regression_score,
                                       std::string* reason) const {
  if (!policy_.circuit_breakers.emergency_disable_file.empty() &&
      std::filesystem::exists(policy_.circuit_breakers.emergency_disable_file)) {
    if (reason) {
      *reason = "autonomy disabled by local emergency switch";
    }
    return true;
  }

  if (CountRecentFailedDeployments() >=
      policy_.circuit_breakers.max_failed_deployments) {
    if (reason) {
      *reason = "circuit breaker open: too many failed deployments";
    }
    return true;
  }

  if (regression_score > policy_.circuit_breakers.max_regression_threshold) {
    if (reason) {
      *reason = "circuit breaker open: regression threshold exceeded";
    }
    return true;
  }

  return false;
}

void PolicyGuard::Audit(ChangeAction action,
                        const std::string& change_id,
                        const std::map<std::string, std::string>& fields) const {
  std::filesystem::create_directories(audit_log_path_.parent_path());

  std::ofstream out(audit_log_path_, std::ios::app);
  if (!out.is_open()) {
    return;
  }

  std::string action_text;
  switch (action) {
    case ChangeAction::Proposed:
      action_text = "proposed";
      break;
    case ChangeAction::Applied:
      action_text = "applied";
      break;
    case ChangeAction::Reverted:
      action_text = "reverted";
      break;
  }

  out << "{\"timestamp\":\"" << ToIso8601Now() << "\",\"action\":\""
      << action_text << "\",\"change_id\":\"" << change_id << "\"";

  for (const auto& [key, value] : fields) {
    out << ",\"" << key << "\":\"" << value << "\"";
  }

  out << "}\n";
}

int PolicyGuard::CountRecentFailedDeployments() const {
  std::ifstream in(audit_log_path_);
  if (!in.is_open()) {
    return 0;
  }

  int failed = 0;
  std::string line;
  while (std::getline(in, line)) {
    if (line.find("\"action\":\"applied\"") != std::string::npos &&
        line.find("\"success\":\"false\"") != std::string::npos) {
      ++failed;
    }
  }

  return failed;
}

}  // namespace game::ai

File: game/ai/policy_guard.py
from __future__ import annotations

import json
import fnmatch
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable, Dict, Iterable, List, Optional

DEFAULT_POLICY_FILE = Path("policies/execution_policy.yaml")
DEFAULT_AUDIT_LOG = Path("logs/autonomy_audit.log")


def _parse_scalar(value: str) -> Any:
    value = value.strip()
    if value in {"true", "True"}:
        return True
    if value in {"false", "False"}:
        return False
    if value.startswith("\"") and value.endswith("\""):
        return value[1:-1]
    if value.startswith("'") and value.endswith("'"):
        return value[1:-1]
    try:
        if "." in value:
            return float(value)
        return int(value)
    except ValueError:
        return value


def _load_simple_yaml(path: Path) -> Dict[str, Any]:
    """Tiny YAML reader for this repository's policy file structure."""
    lines = [line.rstrip("\n") for line in path.read_text(encoding="utf-8").splitlines()]
    root: Dict[str, Any] = {}
    stack: List[tuple[int, Any]] = [(-1, root)]

    for idx, raw in enumerate(lines):
        if not raw.strip() or raw.strip().startswith("#"):
            continue
        indent = len(raw) - len(raw.lstrip(" "))
        text = raw.strip()

        while stack and indent <= stack[-1][0]:
            stack.pop()
        parent = stack[-1][1]

        if text.startswith("- "):
            value = _parse_scalar(text[2:].strip())
            if not isinstance(parent, list):
                raise ValueError(f"invalid list position: {raw}")
            parent.append(value)
            continue

        key, sep, remainder = text.partition(":")
        if not sep:
            raise ValueError(f"invalid yaml line: {raw}")
        key = key.strip()
        remainder = remainder.strip()

        if remainder:
            if isinstance(parent, dict):
                parent[key] = _parse_scalar(remainder)
            else:
                raise ValueError(f"unexpected scalar mapping: {raw}")
            continue

        next_container: Any
        # Determine list vs map by peeking next non-empty line indent/content.
        next_container = {}
        for candidate in lines[idx + 1 :]:
            if not candidate.strip() or candidate.strip().startswith("#"):
                continue
            cand_indent = len(candidate) - len(candidate.lstrip(" "))
            if cand_indent <= indent:
                break
            next_container = [] if candidate.strip().startswith("- ") else {}
            break

        if isinstance(parent, dict):
            parent[key] = next_container
        else:
            raise ValueError(f"unexpected nested mapping: {raw}")
        stack.append((indent, next_container))

    return root


@dataclass
class ResourceRequest:
    cpu_percent: float
    ram_mb: int
    runtime_seconds: int


@dataclass
class EnforcementDecision:
    allowed: bool
    reason: str


class PolicyGuard:
    """Enforces writable path, network, and resource policy for deployments."""

    def __init__(self, policy_file: Path = DEFAULT_POLICY_FILE):
        self.policy_file = policy_file
        self.policy = self._load_policy()

    def _load_policy(self) -> Dict[str, Any]:
        raw = _load_simple_yaml(self.policy_file)
        return raw.get("execution_policy", {})

    def can_write(self, target_paths: Iterable[str]) -> EnforcementDecision:
        writable = self.policy.get("writable_paths", {})
        allow_patterns = writable.get("allow", [])
        deny_patterns = writable.get("deny", [])

        for path in target_paths:
            normalized = path.strip("/")
            if any(fnmatch.fnmatch(normalized, pattern.strip("/")) for pattern in deny_patterns):
                return EnforcementDecision(False, f"path denied by policy: {path}")
            if not any(fnmatch.fnmatch(normalized, pattern.strip("/")) for pattern in allow_patterns):
                return EnforcementDecision(False, f"path not in allowed writable scope: {path}")
        return EnforcementDecision(True, "writable path policy passed")

    def is_network_allowed(self, hosts: Iterable[str]) -> EnforcementDecision:
        network = self.policy.get("network", {})
        allow_hosts = set(network.get("allow", []))
        deny_hosts = set(network.get("deny", []))

        for host in hosts:
            if host in deny_hosts or "*" in deny_hosts:
                if host not in allow_hosts:
                    return EnforcementDecision(False, f"network host denied by policy: {host}")
            if host not in allow_hosts:
                return EnforcementDecision(False, f"network host not allow-listed: {host}")
        return EnforcementDecision(True, "network policy passed")

    def resources_ok(self, request: ResourceRequest) -> EnforcementDecision:
        limits = self.policy.get("resources", {})
        max_cpu = limits.get("max_cpu_percent", 100)
        max_ram = limits.get("max_ram_mb", 16384)
        max_time = limits.get("max_runtime_seconds", 3600)

        if request.cpu_percent > max_cpu:
            return EnforcementDecision(False, f"cpu limit exceeded ({request.cpu_percent}>{max_cpu})")
        if request.ram_mb > max_ram:
            return EnforcementDecision(False, f"ram limit exceeded ({request.ram_mb}>{max_ram})")
        if request.runtime_seconds > max_time:
            return EnforcementDecision(False, f"runtime limit exceeded ({request.runtime_seconds}>{max_time})")
        return EnforcementDecision(True, "resource limits passed")

    def enforce_pre_deploy(
        self,
        target_paths: Iterable[str],
        hosts: Iterable[str],
        request: ResourceRequest,
    ) -> EnforcementDecision:
        for decision in (
            self.can_write(target_paths),
            self.is_network_allowed(hosts),
            self.resources_ok(request),
        ):
            if not decision.allowed:
                return decision
        return EnforcementDecision(True, "all policy checks passed")


class CircuitBreaker:
    """Stops autonomous improvement loop when failure/regression thresholds are exceeded."""

    def __init__(self, policy: Dict[str, Any], state_file: Path = Path("logs/circuit_breaker_state.json")):
        cfg = policy.get("circuit_breakers", {})
        self.max_failed_deployments = int(cfg.get("max_failed_deployments", 3))
        self.max_regression_threshold = float(cfg.get("max_regression_threshold", 0.2))
        self.disable_file = Path(cfg.get("emergency_disable_file", ".autonomy_disabled"))
        self.state_file = state_file
        self.state = self._load_state()

    def _load_state(self) -> Dict[str, Any]:
        if not self.state_file.exists():
            return {"failed_deployments": 0, "last_regression": 0.0}
        with self.state_file.open("r", encoding="utf-8") as handle:
            return json.load(handle)

    def _save_state(self) -> None:
        self.state_file.parent.mkdir(parents=True, exist_ok=True)
        with self.state_file.open("w", encoding="utf-8") as handle:
            json.dump(self.state, handle, indent=2)

    def is_tripped(self) -> EnforcementDecision:
        if self.disable_file.exists():
            return EnforcementDecision(False, "emergency disable file present")
        if int(self.state.get("failed_deployments", 0)) >= self.max_failed_deployments:
            return EnforcementDecision(False, "max failed deployments reached")
        if float(self.state.get("last_regression", 0.0)) > self.max_regression_threshold:
            return EnforcementDecision(False, "regression threshold exceeded")
        return EnforcementDecision(True, "circuit breaker checks passed")

    def record_success(self, regression_score: float = 0.0) -> None:
        self.state["failed_deployments"] = 0
        self.state["last_regression"] = float(regression_score)
        self._save_state()

    def record_failure(self, regression_score: float) -> None:
        self.state["failed_deployments"] = int(self.state.get("failed_deployments", 0)) + 1
        self.state["last_regression"] = float(regression_score)
        self._save_state()


class AutonomyAuditLogger:
    """Appends JSON-line audit entries for proposed/applied/reverted autonomous changes."""

    def __init__(self, audit_log: Path = DEFAULT_AUDIT_LOG):
        self.audit_log = audit_log
        self.audit_log.parent.mkdir(parents=True, exist_ok=True)
        self.audit_log.touch(exist_ok=True)

    def log(self, event: str, patch_id: str, details: Optional[Dict[str, Any]] = None) -> None:
        payload = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "event": event,
            "patch_id": patch_id,
            "details": details or {},
        }
        with self.audit_log.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(payload, sort_keys=True) + "\n")


class PatchDeploymentManager:
    """Evaluates guardrails and applies patch callbacks only when policy permits."""

    def __init__(
        self,
        guard: Optional[PolicyGuard] = None,
        logger: Optional[AutonomyAuditLogger] = None,
    ):
        self.guard = guard or PolicyGuard()
        self.logger = logger or AutonomyAuditLogger()
        self.breaker = CircuitBreaker(self.guard.policy)

    def deploy_patch(
        self,
        patch_id: str,
        target_paths: List[str],
        hosts: List[str],
        request: ResourceRequest,
        regression_score: float,
        apply_fn: Callable[[], bool],
    ) -> EnforcementDecision:
        self.logger.log("proposed", patch_id, {"target_paths": target_paths, "hosts": hosts})

        breaker = self.breaker.is_tripped()
        if not breaker.allowed:
            self.logger.log("reverted", patch_id, {"reason": breaker.reason})
            return breaker

        decision = self.guard.enforce_pre_deploy(target_paths, hosts, request)
        if not decision.allowed:
            self.logger.log("reverted", patch_id, {"reason": decision.reason})
            self.breaker.record_failure(regression_score)
            return decision

        applied = apply_fn()
        if applied:
            self.logger.log("applied", patch_id, {"regression_score": regression_score})
            self.breaker.record_success(regression_score)
            return EnforcementDecision(True, "patch deployed")

        self.logger.log("reverted", patch_id, {"reason": "apply_fn returned false"})
        self.breaker.record_failure(regression_score)
        return EnforcementDecision(False, "deployment callback failed")


def read_recent_audit_entries(limit: int = 20, audit_log: Path = DEFAULT_AUDIT_LOG) -> List[Dict[str, Any]]:
    if not audit_log.exists():
        return []
    with audit_log.open("r", encoding="utf-8") as handle:
        lines = handle.readlines()[-limit:]
    return [json.loads(line) for line in lines if line.strip()]


def audit_summary(audit_log: Path = DEFAULT_AUDIT_LOG) -> Dict[str, int]:
    entries = read_recent_audit_entries(limit=1000, audit_log=audit_log)
    summary = {"proposed": 0, "applied": 0, "reverted": 0}
    for entry in entries:
        event = entry.get("event")
        if event in summary:
            summary[event] += 1
    return summary
